src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingBackgroundService.cs

```diff
@@ -0,0 +1,487 @@
+// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Messaging.EventHubs.Consumer;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+using Azure.Storage.Blobs.Specialized;
+using EnsureThat;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Health.Events.Common;
+using Microsoft.Health.Events.EventCheckpointing;
+using Microsoft.Health.Events.EventConsumers.Service;
+using Microsoft.Health.Events.Telemetry;
+using Microsoft.Health.Events.Telemetry.Exceptions;
+using Microsoft.Health.Logging.Telemetry;
+
+// Description
+// - First, this class gets the number of total processors and total partitions
+// - Next, it will calculate how many partitions should be assigned to each processor
+// - Next, it will lock the partition(s) using blob leases
+// - Next, it will pass the partition id(s) into the AssignedPartitionProcessor which will run until the number of total processors changes
+
+// For a flowchart of the PartitionLockingBackgroundService see:
+// \iomt-fhir\docs\PartitionLocking.md
+
+// todos:
+// - detect if a partition has no ownership updates for some time
+// - handle cases where the ratio of processors to partitions is not divided evenly
+// - detect if acquiring partitions takes too long
+// - replace Task.Delay with something more sophisticated
+
+namespace Microsoft.Health.Events.EventHubProcessor
+{
+    public class PartitionLockingBackgroundService : BackgroundService
+    {
+        private string _processorId;
+
+        private List<int> _ownedPartitions;
+
+        private BlobContainerClient _containerClient;
+
+        private int _totalProcessorsRunning = 0;
+
+        private string[] _eventHubPartitions;
+
+        private EventBatchingService _eventBatchingService;
+
+        private StorageCheckpointClient _checkpointClient;
+
+        private ITelemetryLogger _logger;
+
+        private PartitionLockingBackgroundServiceOptions _partitionLockingOptions;
+
+        private EventHubClientOptions _eventHubClientOptions;
+
+        public PartitionLockingBackgroundService(
+            PartitionLockingBackgroundServiceOptions partitionLockingOptions,
+            EventHubClientOptions eventHubClientOptions,
+            EventBatchingService eventBatchingService,
+            StorageCheckpointClient checkpointClient,
+            ITelemetryLogger logger)
+        {
+            _processorId = Guid.NewGuid().ToString(); // consider using the pod name when running in AKS
+            _ownedPartitions = new List<int>();
```

Comment:
Concurrent Dictionary seems like a better structure.  There are multiple awaited tasks interacting on the collection.  The current order maybe fine but leaves us open to errors as the code changes.  Dictionary would allow you to avoid using contains on the list as well elsewhere in the code.

Reply:
updated